File: CSE216HW4/BinarySearchTree.java

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class BinarySearchTree <T extends Comparable<T>> implements Iterable<T> {

    private Node<T> root;
    private String name;

    public Node<T> getRoot() {
        return root;
    }

    public void setRoot(Node<T> root) {
        this.root = root;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    private static class Node<T> {

        private Node<T> left;
        private Node<T> right;

        private T element;

        public Node(T element){
            this.element = element;
            this.right = null;
            this.left = null;
        }

        public Node(T element, Node<T> left, Node<T> right){
            this.element = element;
            this.right = right;
            this.left = left;
        }

        public T getElement(){
            return this.element;
        }

        public Node<T> getLeft(){
            return this.left;
        }

        public Node<T> getRight(){
            return this.right;
        }


        public void setLeft(Node<T> left) {
            this.left = left;
        }

        public void setRight(Node<T> right) {
            this.right = right;
        }

        public void setElement(T element) {
            this.element = element;
        }
    }

    public BinarySearchTree (Node<T> root, String name){
        this.root = root;
        this.name = name;
    }

    public BinarySearchTree (String name){
        this.name = name;
        this.root = null;
    }


    public void add(Node<T> node){
        if(node == null)
            return;

        if(this.root == null)
            this.root = node;

        addHelper(this.root, node);

    }

    private void addHelper(Node<T> traverse , Node<T> node){
        if(node.element.compareTo(traverse.element) > 0){
            if(traverse.right == null){
                traverse.right = node;
            } else {
                addHelper(traverse.right, node);
            }
        }

        else if(node.element.compareTo(traverse.element) < 0){
            if(traverse.left == null){
                traverse.left = node;
            } else {
                addHelper(traverse.left, node);
            }
        } else {
            return;
        }
    }

    public void addAll(List<T> list){
        for(T element : list)
            add(new Node<T>(element));
    }

    private void printInorder(Node<T> root){                     //CHANGE THIS METHOD BACK TO PRIVATE
        //System.out.println(this.root.element + " root");

        if(root.left != null)
            printInorder(root.left);

        System.out.print(root.element + " ");

        if(root.right != null)
            printInorder(root.right);
    }

    private ArrayList<Node<T>> inorder(Node<T> root, ArrayList<Node<T>> list){

        if(root.left != null)
            inorder(root.left, list);

        list.add(root);

        if(root.right != null)
            inorder(root.right, list);

        return list;
    }

    private String inorderString(Node<T> root, String concat){
        concat += root.element + "";

        if(root.left != null){
            //concat = concat + "L:";
            concat = inorderString(root.left, concat + " L:(");
            concat += ")";
        }

        if(root.right != null){
            //concat = concat + "R:";
            concat = inorderString(root.right, concat + " R:(");
            concat += ")";
        }

        return concat;
    }

    @Override
    public String toString(){
        ArrayList<Node<T>> list = inorder(this.root, new ArrayList<>());

        if(this.root == null)
            return "null BST";

        String toReturn = "[" + this.name + "] " + inorderString(this.root, "");

        return toReturn;

    }


    @Override
    public Iterator<T> iterator() {
        Iterator<T> iterator = new Iterator<T>() {
            List<Node<T>> BST = inorder(root, new ArrayList<>());

            @Override
            public boolean hasNext() {
                return BST.size() != 0;
            }

            @Override
            public T next() {
                return BST.remove(0).element;
            }

            @Override
            public void remove(){
                BST.remove(0);
            }
        };

        return iterator;
    }

    private static class MergeThread<T extends Comparable<T>> implements Runnable{

        private BinarySearchTree<T> tree1;
        private BinarySearchTree<T> merged;

        public MergeThread(BinarySearchTree<T> t1, BinarySearchTree<T> mergeTree){
            this.tree1 = t1;
            this.merged = mergeTree;
        }



        @Override
        public void run() {

            for(T element : tree1){
                merged.add(new Node<>(element));
            }

        }
    }

    public static <T extends Comparable<T>> List<T> merge (BinarySearchTree<T> t1, BinarySearchTree<T> t2){
        BinarySearchTree<T> mergedTree = new BinarySearchTree<>("Merged Tree");

        Thread t1Thread = new Thread(new MergeThread<T>(t1,mergedTree));
        Thread t2Thread = new Thread(new MergeThread<T>(t2,mergedTree));

        t1Thread.start();
        t2Thread.start();

        while(true){
            if(!(t1Thread.isAlive() || t2Thread.isAlive())) {
                return BinarySearchTree.nodeToElementList(mergedTree.inorder(mergedTree.root, new ArrayList<>()));
            }
        }
    }

    private static <T> List<T> nodeToElementList(List<Node<T>> list){
        List<T> toReturn = new ArrayList<>();
        for(Node<T> node : list){
            toReturn.add(node.element);
        }
        return  toReturn;
    }

}

File: CSE216HW4/binarysearchtree.py
class Node:
    element = None
    left = None
    right = None

    def __init__ (self):
        self.element = None

    def setData(self, element):
        self.element = element
    
    def setLeft(self, leftNode):
        self.left = leftNode

    def setRight(self, rightNode):
        self.right = rightNode

    def inorderNode(root , list):
        if(root.left != None):
            Node.inorderNode(root.left, list)
        
        list.append(root)

        if(root.right != None):
            Node.inorderNode(root.right,list)

        return list


    def __iter__(self):
        node = self
        listToIterate = Node.inorderNode(root = node,list=[])

        for element in listToIterate:
            yield element.element

        return

    def __str__(self):
        return str(self.element)

class BinarySearchTree:
    root = None
    name = None


    def __init__(self, name, root):
        self.name = name
        self.root = root

    def setRoot(self, root):
        self.root = root

    def add(self, node):
        ## TREE DOES NOT TAKE DUPLICATE VALUES - NOT SPECIFIED IN INSTRUCTIONS
        if(node == None):
            return
        
        if(self.root.element == None):
            self.root = node


        self.addHelper(self.root, node)

    
    def addHelper(self, traverse, node):
        ## TREE DOES NOT TAKE DUPLICATE VALUES - NOT SPECIFIED IN INSTRUCTIONS
        if(node.element < traverse.element):
            if(traverse.left == None):
                traverse.left = node
            else:
                self.addHelper(traverse.left, node)

        if(node.element > traverse.element):
            if(traverse.right == None):
                traverse.right = node
            else:
                self.addHelper(traverse.right, node)

    
    def add_all(self, *list):
        ## TREE DOES NOT TAKE DUPLICATE VALUES - NOT SPECIFIED IN INSTRUCTIONS
        for element in list:
            node = Node()
            node.element = element
            self.add(node)

        return

    def inorder(self, root, list):
        if(root.left != None):
            self.inorder(root.left, list)

        list.append(root)

        if(root.right != None):
            self.inorder(root.right, list)

        return list


    def inorderElements(self, root, list):
        if(root.left != None):
            self.inorderElements(root.left, list)

        list.append(root.element)

        if(root.right != None):
            self.inorderElements(root.right, list)

        return list


    def inorderString(self, node, concat):
        
        concat = "" + concat + str(node.element) + ""

        if(node.left != None):
            concat = self.inorderString(node.left, concat + "L:(")
            concat += ")"
        
        if(node.right != None):
            concat = self.inorderString(node.right, concat + "R:(")
            concat += ")"

        return concat


    def __str__(self):
        list = self.inorder(self.root, [])

        if(self.root == None):
            return "Null bruh"

        toReturn = "[" + self.name + "] " + self.inorderString(self.root, "")

        return toReturn

    def __iter__(self):
        listToIterate = self.inorder(self.root,[])

        for node in listToIterate:
            yield node.element

        return
        



class Merger:
    def __init__(self):
        return

    
    def merge(t1, t2):
        mergedBST = BinarySearchTree("Merged Tree", root = Node())
        t1Iterator = t1.__iter__()
        t2Iterator = t2.__iter__()

        t1Done = False
        t2Done = False

        while(not t1Done or not t2Done):
            try:
                element = next(t1Iterator)
                node = Node()
                node.element = element
                mergedBST.add(node)
            except StopIteration:
                t1Done = True

            try:
                element = next(t2Iterator)
                node = Node()
                node.element = element
                mergedBST.add(node)
            except StopIteration:
                t2Done = True

        return mergedBST.inorderElements(mergedBST.root,[])


if __name__ == "__main__":
    tree = BinarySearchTree(name="Oak", root = Node())
    tree.add_all(5, 3, 9, 0) # adds the elements in the order 5, 3, 9, and then 0
    print(tree)
    

    t1 = BinarySearchTree(name="Oak", root=Node())
    t2 = BinarySearchTree(name="Birch", root=Node())
    t1.add_all(5, 3, 9, 0)
    t2.add_all(1, 0, 10, 2, 7)

    for x in t1:
        print(x)


    for x in t2.root:
        print(x)

    print(Merger.merge(t1, t2))